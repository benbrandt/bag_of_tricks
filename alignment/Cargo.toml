[package]
name = "alignment"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2018"
resolver = "2"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
rand = "0.8"
rand_pcg = "0.3"
serde = { version = "1", features = ["derive"] }
strum = { version = "0.20", features = ["derive"] }
