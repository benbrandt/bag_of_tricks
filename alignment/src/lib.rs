#![deny(clippy::all)]
#![warn(clippy::pedantic)]

use std::{convert::TryFrom, f64::consts::E, fmt};

use rand::{prelude::SliceRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::{Display, EnumIter, IntoEnumIterator};

#[derive(Clone, Copy, Deserialize, Display, EnumIter, PartialEq, Serialize)]
pub enum Attitude {
    Chaotic,
    Lawful,
    Neutral,
}

#[derive(Clone, Copy, Deserialize, Display, EnumIter, PartialEq, Serialize)]
pub enum Morality {
    Evil,
    Good,
    Neutral,
}

/// Choose from a given list of options weighted by number of influences for a given type
fn weighted_choice<T, I>(rng: &mut impl Rng, options: I, influences: &[T]) -> T
where
    T: Clone + Copy + PartialEq,
    I: Clone + Iterator<Item = T>,
{
    *options
        .collect::<Vec<T>>()
        .choose_weighted(rng, |t| {
            E.powi(i32::try_from(influences.iter().filter(|i| t == *i).count()).unwrap_or(0))
        })
        .unwrap()
}

/// Character alignment, both attitude and morality
#[derive(Deserialize, Serialize)]
pub struct Alignment(pub Attitude, pub Morality);

impl Alignment {
    /// Generate alignment, weighted by influences from other choices on the character sheet
    pub fn gen(
        rng: &mut impl Rng,
        attitude_influences: &[Attitude],
        morality_influences: &[Morality],
    ) -> Self {
        let attitude = weighted_choice(rng, Attitude::iter(), attitude_influences);
        let morality = weighted_choice(rng, Morality::iter(), morality_influences);
        Self(attitude, morality)
    }
}

impl fmt::Display for Alignment {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self(Attitude::Neutral, Morality::Neutral) => write!(f, "Neutral"),
            Self(a, m) => write!(f, "{} {}", a, m),
        }
    }
}

/// Trait to describe how this entity affects a character's alignment
pub trait AlignmentInfluences {
    /// List of attitude influences
    fn attitude(&self) -> Vec<Attitude> {
        vec![]
    }

    /// List of morality influences
    fn morality(&self) -> Vec<Morality> {
        vec![]
    }
}
