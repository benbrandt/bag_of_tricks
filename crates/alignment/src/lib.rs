#![deny(clippy::all)]
#![warn(clippy::pedantic)]

use std::{convert::TryFrom, f64::consts::E, fmt};

use rand::{prelude::SliceRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::{Display, EnumIter, IntoEnumIterator};

fn exp_weight(val: usize) -> f64 {
    E.powi(i32::try_from(val).unwrap_or_default())
}

#[derive(Clone, Copy, Deserialize, Display, EnumIter, Eq, PartialEq, Serialize)]
pub enum Attitude {
    Chaotic,
    Lawful,
    Neutral,
}

impl Attitude {
    fn weight(self, influences: &[Self]) -> f64 {
        exp_weight(influences.iter().filter(|&i| i == &self).count())
    }
}

#[derive(Clone, Copy, Deserialize, Display, EnumIter, Eq, PartialEq, Serialize)]
pub enum Morality {
    Evil,
    Good,
    Neutral,
}

impl Morality {
    fn weight(self, influences: &[Self]) -> f64 {
        exp_weight(influences.iter().filter(|&i| i == &self).count())
    }
}

/// Trait to describe how this entity affects a character's alignment
pub trait AlignmentInfluences {
    /// List of attitude influences
    fn attitude(&self) -> Vec<Attitude> {
        vec![]
    }

    /// List of morality influences
    fn morality(&self) -> Vec<Morality> {
        vec![]
    }
}

/// Character alignment, both attitude and morality
#[derive(Clone, Deserialize, Serialize)]
pub struct Alignment(pub Attitude, pub Morality);

impl Alignment {
    /// Generate alignment, weighted by influences from other choices on the character sheet
    ///
    /// # Panics
    ///
    /// Will panic if weighting logic is wrong or there are no attitudes/moralities at all
    pub fn gen(
        rng: &mut impl Rng,
        attitude_influences: &[Attitude],
        morality_influences: &[Morality],
    ) -> Self {
        let attitude = *Attitude::iter()
            .collect::<Vec<_>>()
            .choose_weighted(rng, |a| a.weight(attitude_influences))
            .unwrap();
        let morality = *Morality::iter()
            .collect::<Vec<_>>()
            .choose_weighted(rng, |a| a.weight(morality_influences))
            .unwrap();
        Self(attitude, morality)
    }

    /// Weight of a particular alignment based on influences.
    /// Useful for comparing things like deities.
    #[must_use]
    pub fn weight(
        &self,
        attitude_influences: &[Attitude],
        morality_influences: &[Morality],
    ) -> f64 {
        self.0.weight(attitude_influences) + self.1.weight(morality_influences)
    }
}

impl AlignmentInfluences for Alignment {
    fn attitude(&self) -> Vec<Attitude> {
        vec![self.0]
    }

    fn morality(&self) -> Vec<Morality> {
        vec![self.1]
    }
}

impl fmt::Display for Alignment {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self(Attitude::Neutral, Morality::Neutral) => write!(f, "Neutral"),
            Self(a, m) => write!(f, "{} {}", a, m),
        }
    }
}
