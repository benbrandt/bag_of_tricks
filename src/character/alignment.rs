use std::fmt;

use rand::{prelude::IteratorRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::IntoEnumIterator;
use strum_macros::{Display, EnumIter};

#[derive(Deserialize, Display, EnumIter, Serialize)]
enum Attitude {
    Chaotic,
    Lawful,
    Neutral,
}

#[derive(Deserialize, Display, EnumIter, Serialize)]
enum Morality {
    Evil,
    Good,
    Neutral,
}

#[derive(Deserialize, Serialize)]
pub(crate) struct Alignment(Attitude, Morality);

impl Alignment {
    pub(crate) fn gen(rng: &mut impl Rng) -> Self {
        let attitude = Attitude::iter().choose(rng).unwrap();
        let morality = Morality::iter().choose(rng).unwrap();
        Self(attitude, morality)
    }
}

impl fmt::Display for Alignment {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self(Attitude::Neutral, Morality::Neutral) => write!(f, "Neutral"),
            Self(a, m) => write!(f, "{} {}", a, m),
        }
    }
}
