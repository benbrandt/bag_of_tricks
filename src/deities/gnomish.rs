use crate::alignment::{Alignment, Attitude, Morality};

use super::{Deity, Domain, Pantheon};

pub(crate) struct Gnomish;

impl Pantheon for Gnomish {
    fn deities() -> Vec<Deity> {
        vec![
            Deity {
                name: "Baervan Wildwanderer",
                titles: vec!["god of woodlands"],
                alignment: Alignment(Attitude::Neutral, Morality::Good),
                domains: vec![Domain::Nature],
                symbols: vec!["Face of a raccoon"],
            },
            Deity {
                name: "Baravar Cloakshadow",
                titles: vec!["god of illusion and deception"],
                alignment: Alignment(Attitude::Neutral, Morality::Good),
                domains: vec![Domain::Arcana, Domain::Trickery],
                symbols: vec!["Dagger against a hooded cloak"],
            },
            Deity {
                name: "Callarduran Smoothhands",
                titles: vec!["god of mining and carving stone"],
                alignment: Alignment(Attitude::Neutral, Morality::Neutral),
                domains: vec![Domain::Knowledge, Domain::Nature],
                symbols: vec!["Golden signet ring wiht a six-pointed star"],
            },
            Deity {
                name: "Flandal Steelskin",
                titles: vec!["god of metalwork"],
                alignment: Alignment(Attitude::Neutral, Morality::Good),
                domains: vec![Domain::Knowledge],
                symbols: vec!["Flaming hammer"],
            },
            Deity {
                name: "Gaerdal Ironhand",
                titles: vec!["god of protection"],
                alignment: Alignment(Attitude::Lawful, Morality::Good),
                domains: vec![Domain::War],
                symbols: vec!["Iron band"],
            },
            Deity {
                name: "Garl Glittergold",
                titles: vec!["god of trickery and gems", "god of trickery and wiles"],
                alignment: Alignment(Attitude::Lawful, Morality::Good),
                domains: vec![Domain::Trickery],
                symbols: vec!["Gold nugget"],
            },
            Deity {
                name: "Nebelun",
                titles: vec!["god of invention and luck"],
                alignment: Alignment(Attitude::Chaotic, Morality::Good),
                domains: vec![Domain::Knowledge, Domain::Trickery],
                symbols: vec!["Bellows and a lizard tail"],
            },
            Deity {
                name: "Segojan Earthcaller",
                titles: vec!["god of earth and the dead"],
                alignment: Alignment(Attitude::Neutral, Morality::Good),
                domains: vec![Domain::Light],
                symbols: vec!["Glowing gemstone"],
            },
            Deity {
                name: "Urdlen",
                titles: vec!["god of greed and murder"],
                alignment: Alignment(Attitude::Chaotic, Morality::Evil),
                domains: vec![Domain::Death, Domain::War],
                symbols: vec!["White clawed mole emerging from ground"],
            },
        ]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_deities() {
        insta::assert_yaml_snapshot!(Gnomish::deities());
    }
}
