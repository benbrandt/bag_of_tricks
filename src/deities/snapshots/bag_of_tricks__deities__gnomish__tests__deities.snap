---
source: src/deities/gnomish.rs
expression: "Gnomish::deities()"

---
- name: Baervan Wildwanderer
  titles:
    - god of woodlands
  alignment:
    - Neutral
    - Good
  domains:
    - Nature
  symbols:
    - Face of a raccoon
- name: Baravar Cloakshadow
  titles:
    - god of illusion and deception
  alignment:
    - Neutral
    - Good
  domains:
    - Arcana
    - Trickery
  symbols:
    - Dagger against a hooded cloak
- name: Callarduran Smoothhands
  titles:
    - god of mining and carving stone
  alignment:
    - Neutral
    - Neutral
  domains:
    - Knowledge
    - Nature
  symbols:
    - Golden signet ring wiht a six-pointed star
- name: Flandal Steelskin
  titles:
    - god of metalwork
  alignment:
    - Neutral
    - Good
  domains:
    - Knowledge
  symbols:
    - Flaming hammer
- name: Gaerdal Ironhand
  titles:
    - god of protection
  alignment:
    - Lawful
    - Good
  domains:
    - War
  symbols:
    - Iron band
- name: Garl Glittergold
  titles:
    - god of trickery and gems
    - god of trickery and wiles
  alignment:
    - Lawful
    - Good
  domains:
    - Trickery
  symbols:
    - Gold nugget
- name: Nebelun
  titles:
    - god of invention and luck
  alignment:
    - Chaotic
    - Good
  domains:
    - Knowledge
    - Trickery
  symbols:
    - Bellows and a lizard tail
- name: Segojan Earthcaller
  titles:
    - god of earth and the dead
  alignment:
    - Neutral
    - Good
  domains:
    - Light
  symbols:
    - Glowing gemstone
- name: Urdlen
  titles:
    - god of greed and murder
  alignment:
    - Chaotic
    - Evil
  domains:
    - Death
    - War
  symbols:
    - White clawed mole emerging from ground

